retro_data_structures/formats/mrea.py

```diff
@@ -711,45 +797,91 @@ def build_mlvl_dependencies(self, only_modified: bool = False):
                     continue
 
                 layer = self.get_layer(layer_name)
-                if only_modified and not layer.is_modified:
+                if only_modified and not layer.is_modified():
                     continue
 
                 layer_deps[layer.index].extend(missing)
 
-        layer_deps = [
-            [dep for dep in layer if not dep.exclude_for_mlvl]
-            for layer in layer_deps
+        if only_modified:
+            # assume we never modify these
+            non_layer = self.dependencies.non_layer
+        else:
+            non_layer = list(self.build_non_layer_dependencies())
+            if "!!non_layer!!" in _hardcoded_dependencies.get(self.mrea_asset_id, {}):
+                non_layer.extend(_hardcoded_dependencies[self.mrea_asset_id]["!!non_layer!!"])
+
+        self.dependencies = AreaDependencies(layer_deps, non_layer)
+
+    def ensure_dependencies_in_paks(self):
+        asset_manager = self._parent_mlvl.asset_manager
+        paks = asset_manager.find_paks(self.mrea_asset_id)
+        for dep in self.dependencies.all_dependencies:
+            for pak in paks:
+                asset_manager.ensure_present(pak, dep.id)
+
+    def build_module_dependencies(self, only_modified: bool = False):
+        if self._parent_mlvl.asset_manager.target_game == Game.PRIME:
+            return
+        layers = list(self.layers)
+        layer_rels = [
+            list(
+                layer.build_module_dependencies()
+                if (not only_modified) or layer.is_modified() else
+                layer.module_dependencies
+            ) for layer in layers
         ]
 
-        offset = 0
-        offsets = []
-        for layer_dep in layer_deps:
-            offsets.append(offset)
-            offset += len(layer_dep)
+        for instance in self.generated_objects_layer.instances:
+            layer = instance.id.layer
+            if (
+                (not only_modified)
+                or layers[layer].is_modified()
+            ):
+                layer_rels[layer].extend(instance.get_properties().modules())
```

Comment:
`modules` is a class method, this can be `instance.type.modules()` and be faster